### PR TITLE
8284780: Need methods to create pre-sized HashSet and LinkedHashSet

### PR DESCRIPTION
**Pull request by XenoAmess**
as title.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8284780](https://bugs.openjdk.org/browse/JDK-8284780): Need methods to create pre-sized HashSet and LinkedHashSet
 * [JDK-8287419](https://bugs.openjdk.org/browse/JDK-8287419): Need methods to create pre-sized HashSet and LinkedHashSet (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [98bfb0e1](https://git.openjdk.java.net/jdk/pull/8302/files/98bfb0e1e3a191aed70ce9284cfbb63f9e8d3667)
 * [Brian Burkhalter](https://openjdk.java.net/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [98bfb0e1](https://git.openjdk.java.net/jdk/pull/8302/files/98bfb0e1e3a191aed70ce9284cfbb63f9e8d3667)
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to [98bfb0e1](https://git.openjdk.java.net/jdk/pull/8302/files/98bfb0e1e3a191aed70ce9284cfbb63f9e8d3667)
 * [Anthony Scarpino](https://openjdk.java.net/census#ascarpino) (@ascarpino - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8302/head:pull/8302` \
`$ git checkout pull/8302`

Update a local copy of the PR: \
`$ git checkout pull/8302` \
`$ git pull https://git.openjdk.java.net/jdk pull/8302/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8302`

View PR using the GUI difftool: \
`$ git pr show -t 8302`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8302.diff">https://git.openjdk.java.net/jdk/pull/8302.diff</a>

</details>

### CODE REVIEW

**Comment by liach** [comment]
Need to add apiNote documentation section to capacity-based constructors like for maps.

**Comment by XenoAmess** [comment]
> /csr needed

@stuart-marks csr please

**Comment by XenoAmess** [comment]
> Need to add apiNote documentation section to capacity-based constructors like for maps.

@liach done.

**Comment by XenoAmess** [comment]
> suggest keeping the changes under src in java.base, java.net.http, java.rmi, and jdk.zipfs, and omitting all the other changes. Also keep the changes under test/jdk.

done.

File: test/jdk/java/util/HashMap/WhiteBoxResizeTest.java
Comment (ExE-Boss) [inline test/jdk/java/util/HashMap/WhiteBoxResizeTest.java:360]: These probably need a `mapField.setAccessible(true)` call, or a `VarHandle` for the `HashSet.map` field.

File: test/jdk/java/util/HashMap/WhiteBoxResizeTest.java
Comment (liach) [inline test/jdk/java/util/HashMap/WhiteBoxResizeTest.java:354]: For clarity, consider using var handle/method handle and keep the reflection code in the `WhiteBoxResizeTest` constructor.

File: src/java.base/share/classes/java/util/HashSet.java
Comment (stuart-marks) [inline src/java.base/share/classes/java/util/HashSet.java:398]: Please use "elements" instead of "items" throughout for the objects that are members of the HashSet. This includes the API notes above as well as the specs and API notes in LinkedHashSet.

File: test/jdk/java/util/HashMap/WhiteBoxResizeTest.java
Comment (stuart-marks) [reply test/jdk/java/util/HashMap/WhiteBoxResizeTest.java:360]: Yes, this test fails with IllegalAccessException. Probably it's easiest to use a VarHandle to get private fields, similar to other usage already in this test.

This test case is a bit odd though in that it's supposed to test HashSet and LinkedHashSet but it mostly actually tests HashMap. It creates the Set instance and immediately extracts the HashMap, which is then passed to the actual test, which operates directly on the HashMap. It would be preferable to create a Set; add an element (so that it's properly allocated); and then make assertions over the Set (which involve extracting the HashMap, etc.) It seems like there should be factoring that allows this sort of arrangement to be retrofitted without adding too much complication.

Finally, please add "8284780" to the `@bug` line at the top of this test.

**Review by stuart-marks — COMMENTED** [review]
I looked at all the use sites and they look fine. Some look like they could use additional cleanup, but that's probably beyond the scope of this change. (Also, I haven't seen `StringTokenizer` in a long time....) It's amazing how many bugs there are -- the majority look like they allocated the HashSet with the wrong capacity! Again, this proves the worth of these new APIs.

**Comment by naotoj** [comment]
> (Also, I haven't seen `StringTokenizer` in a long time....)

That's some old code lingering in locale-related stuff. Will fix them after this PR gets integrated.

File: test/jdk/java/util/HashMap/WhiteBoxResizeTest.java
Comment (XenoAmess) [reply test/jdk/java/util/HashMap/WhiteBoxResizeTest.java:360]: > Yes, this test fails with IllegalAccessException. Probably it's easiest to use a VarHandle to get private fields, similar to other usage already in this test.
> 
> This test case is a bit odd though in that it's supposed to test HashSet and LinkedHashSet but it mostly actually tests HashMap. It creates the Set instance and immediately extracts the HashMap, which is then passed to the actual test, which operates directly on the HashMap.

Considering about it being a whitebox test, it is not very weird, but still weird enough.
A better way might be wrap it into another layer.
Will try it when I have time.

File: src/java.base/share/classes/java/util/HashSet.java
Comment (XenoAmess) [reply src/java.base/share/classes/java/util/HashSet.java:398]: @stuart-marks done.

File: test/jdk/java/util/HashMap/WhiteBoxResizeTest.java
Comment (XenoAmess) [reply test/jdk/java/util/HashMap/WhiteBoxResizeTest.java]: @liach I refactored the tests. please have a look.

File: test/jdk/java/util/HashMap/WhiteBoxResizeTest.java
Comment (XenoAmess) [reply test/jdk/java/util/HashMap/WhiteBoxResizeTest.java]: @stuart-marks I refactored the tests. please have a look.

Comment (XenoAmess) [reply test/jdk/java/util/HashMap/WhiteBoxResizeTest.java]: @stuart-marks 8284780 added.

**Comment by XenoAmess** [comment]
@stuart-marks @liach 
Hi guys.
After a good sleep and a rethink, I re-wrote some part of the test. please have a look.
It seems be logically better than the way before.

File: src/java.base/share/classes/java/util/HashSet.java
Comment (liach) [inline src/java.base/share/classes/java/util/HashSet.java:391]:
```suggestion
     * @param <E> the type of elements maintained by this set
```

File: src/java.base/share/classes/java/util/HashSet.java
Comment (liach) [inline src/java.base/share/classes/java/util/HashSet.java:130]:
```suggestion
     * an expected number of elements, use {@link #newHashSet(int) newHashSet}.
```

File: src/java.base/share/classes/java/util/HashSet.java
Comment (liach) [inline src/java.base/share/classes/java/util/HashSet.java:147]:
```suggestion
     * an expected number of elements, use {@link #newHashSet(int) newHashSet}.
```

File: src/java.base/share/classes/java/util/LinkedHashSet.java
Comment (liach) [inline src/java.base/share/classes/java/util/LinkedHashSet.java:131]:
```suggestion
     * an expected number of elements, use {@link #newLinkedHashSet(int) newLinkedHashSet}.
```

File: src/java.base/share/classes/java/util/LinkedHashSet.java
Comment (liach) [inline src/java.base/share/classes/java/util/LinkedHashSet.java:148]:
```suggestion
     * an expected number of elements, use {@link #newLinkedHashSet(int) newLinkedHashSet}.
```

File: src/java.base/share/classes/java/util/LinkedHashSet.java
Comment (liach) [inline src/java.base/share/classes/java/util/LinkedHashSet.java:212]:
```suggestion
     * @param <E> the type of elements maintained by this set
```

**Comment by stuart-marks** [comment]
Thanks for the updates. I've made a couple minor changes to the specs; please merge the latest commit from this branch:

https://github.com/stuart-marks/jdk/tree/pull/8302

I've created a CSR and have included these changes in it. Please review: [JDK-8287419](https://bugs.openjdk.java.net/browse/JDK-8287419)

I'll look at the test changes later. I wanted to get the CSR moving first, since it will take a few days (and a long weekend in the US is coming up).

**Comment by XenoAmess** [comment]
@stuart-marks @liach done.

File: src/java.base/share/classes/sun/util/cldr/CLDRLocaleProviderAdapter.java
Comment (naotoj) [inline src/java.base/share/classes/sun/util/cldr/CLDRLocaleProviderAdapter.java:227]: Hi @XenoAmess, during the refactoring of the above `StringTokenizer`, it turned out that this `HashSet` can be replaced with `Set.of()`. So you can leave this piece as it is, as I will take care of it with JDK-8287340.

File: src/java.base/share/classes/sun/util/locale/provider/JRELocaleProviderAdapter.java
Comment (naotoj) [inline src/java.base/share/classes/sun/util/locale/provider/JRELocaleProviderAdapter.java:461]: Same as above.

File: src/java.base/share/classes/sun/util/cldr/CLDRLocaleProviderAdapter.java
Comment (XenoAmess) [reply src/java.base/share/classes/sun/util/cldr/CLDRLocaleProviderAdapter.java]: @naotoj done.

File: src/java.base/share/classes/sun/util/locale/provider/JRELocaleProviderAdapter.java
Comment (XenoAmess) [reply src/java.base/share/classes/sun/util/locale/provider/JRELocaleProviderAdapter.java]: @naotoj done.

**Comment by stuart-marks** [comment]
Reviewers for i18n, net, nio, and security, please review call site changes in your areas. Thanks.

File: src/java.base/share/classes/java/util/Calendar.java
Comment (naotoj) [inline src/java.base/share/classes/java/util/Calendar.java:2648]: This can be replaced with `SET = Set.of("gregory", "buddhist", "japanese");`.

**Review by naotoj — APPROVED** [review]
Reviewed i18n-related changes and they look good. One minor suggestion in `Calendar`, but that can be applied later.

**Review by bplb — APPROVED** [review]
`java.io` and `java.nio` look all right.

**Review by dfuch — APPROVED** [review]
Changes to `net` and `http` look good.

File: test/jdk/java/util/HashMap/WhiteBoxResizeTest.java
Comment (XenoAmess) [reply test/jdk/java/util/HashMap/WhiteBoxResizeTest.java:428]: @stuart-marks done.

File: src/java.base/share/classes/java/util/Calendar.java
Comment (XenoAmess) [reply src/java.base/share/classes/java/util/Calendar.java]: @naotoj Yes it can. I did a further clean up to it, please have a look.

**Comment by stuart-marks** [comment]
Running tests and awaiting review from security team. Our internal test system is backlogged and tests might not complete in time to get into JDK 19.

**Review by ascarpino — APPROVED** [review]
I gave a quick look at the security files touched and seems straightforward. I didn't see any problems

**Comment by stuart-marks** [comment]
@liach 

> 'the new' fix should be applied to newHashMap etc. too.

I've filed [JDK-8289872](https://bugs.openjdk.org/browse/JDK-8289872) for this. It's a small thing but best to fix it before we forget about it.

**Comment by XenoAmess** [comment]
> @liach
> 
> > 'the new' fix should be applied to newHashMap etc. too.
> 
> I've filed [JDK-8289872](https://bugs.openjdk.org/browse/JDK-8289872) for this. It's a small thing but best to fix it before we forget about it.

Sorry for I already forgotten it.

**Pull request closed without merging**
